### PR TITLE
WT-16983 Mark disaggregated btrees outdated at step-down

### PR DESCRIPTION
**Pull request by dhly-etc**
At step-down a leader marks its disaggregated btrees readonly so the follower does not checkpoint them. A handle left readonly across the following step-up is skipped by checkpoint, so its disaggregated block-size accounting drifts from the metadata and a later checkpoint trips the ckpt->size == __wt_block_disagg_get_size assertion.

Also mark the handles outdated at step-down. The next leader then opens a fresh handle for each tree, which is checkpointed normally as leader, keeping the block-size accounting consistent.

Add test_layered_stepup10, which runs two step-down/step-up cycles and asserts the accounting stays consistent.

### CODE REVIEW

**Comment by luke-pearson** [comment]
This fix is framed as a checkpoint size related fix, but isn't this a much bigger issue where btrees aren't marked as writeable? Shouldn't that have more impact?

**Comment by dhly-etc** [comment]
As an aside, including the added test in the code coverage does yield 100% coverage of the new code, but I wasn't sure about the policy on what gets included there. I'm happy to add it if we're generally liberal in adding new tests for coverage.

@luke-pearson Yes, there's a much bigger impact here around step up/down/up cycles, but since we currently crash on step down, it's not really ever exercised currently. I encountered this issue in testing around parallelizing ingest drain, and figured it might be worth fixing as-is since it seems self-contained.

**Comment by quchenhao** [comment]
Sorry I misunderstood the intention of the change so it is to fix a checkpoint on a btree is missed in a step-down step-up sequence? We have changed step-down to be a restart? Don't we open the btree again when we step up? If we don't do that, instead of this change, I think we should also mark the live btrees as WT_DHANDLE_OUTDATED in step-down to ensure we reopen the btree from the latest checkpoint during step-up.

File: src/conn/conn_layered.c [hunk 1]
```diff
@@ -1615,6 +1615,42 @@ __disagg_restart_checkpoint(WT_SESSION_IMPL *session)
...
+            break;
+
+        /* Only care about open disaggregated btree dhandles. */
+        if (!WT_DHANDLE_BTREE(dhandle) || !F_ISSET(dhandle, WT_DHANDLE_OPEN))
+            continue;
+
+        btree = (WT_BTREE *)dhandle->handle;
+
+        if (!F_ISSET(btree, WT_BTREE_DISAGGREGATED) || !F_ISSET(btree, WT_BTREE_READONLY))
+            continue;
```
Comment (quchenhao) [inline src/conn/conn_layered.c]: We cannot clear the read only flag from the checkpoint btree opened in the follower mode.

**Comment by quchenhao** [comment]
I would suggest we mark the dhandle as WT_DHANDLE_OUTDATED in __disagg_mark_btrees_readonly_then_step_down

**Comment by ivankochin** [comment]
If my understanding of the current logic is correct, we always open a dhandle in one of two forms:

* a live btree (we open a dhandle for table_name.wt_stable), or
* a checkpointed btree (we open a dhandle for table_name.wt_stable/WiredTigerCheckpoint#).

When we are on a follower, we should always open the stable table using a checkpointed view, and this btree is always read-only. I don’t think we ever open a checkpointed btree on the leader.

So, theoretically (since step-down is currently not working), when we step down we may have a bunch of live btrees that we mark as read-only, but that sounds more like protection against accidental writes, since we should always reopen all btrees on a specific checkpoint.

Given that, on step-up, even if we still have some checkpointed btrees open, I don’t think we should remove the read-only flag from them. Instead, we should only remove it from the live btrees that were marked read-only during step-down.

Another thing is that step-down is not currently fully supported. We have plans to support it before public preview, so I am wondering whether it would make sense to implement this change together with the step-down work so that we get better testing coverage. @Siddhartha8899 is going to lead the project, so he might be interested in this change as well.

**Comment by dhly-etc** [comment]
Point well-taken that clearing the read-only flag is a bad solution here (also confirmed via further testing that this could cause assertion failures in eviction 😓). The WT_DHANDLE_OUTDATED suggestion appears to be a good one, as it prevents the same assertions failures by forcing a reopen.

As to the broader question Ivan raised about whether to fix this now or defer to the step-down project, it's a fair one. As I mentioned, I ran into this while trying to test some of the edge cases in my POC for parallel ingest drain (WT-17449). In theory we can decouple these by simply saying "step-down not implemented yet" consistently across both tickets, and leave any clean-up to the step-down project. But I think making a good-faith attempt to add basic step-down/step-up cycle testing in a few places can help prevent us from adding tech debt in other tickets that Sid would have to pay down. 😅

**Review by quchenhao — APPROVED** [review]
lgtm

**Comment by Siddhartha8899** [comment]
Thanks for tagging me on this one, @ivankochin.

@dhly-etc, for some context: I've been investigating whether elegant stepdown works in WT testing, that is, whether we can avoid crashing or restarting and instead simply call `reconfigure(role=follower)`, especially in [test/format](https://github.com/wiredtiger/wiredtiger/blob/develop/test/format/format_disagg.c#L244). During that work, I ran into consistent failures and developed a theory that some of them could be addressed by marking dhandles as outdated during stepdown.

When I saw this change, I couldn't help but wonder whether it might have resolved the remaining issues related to elegant stepdown. I ran a [patch](https://spruce.corp.mongodb.com/version/6a20c12014fb5900070c0a3a/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC) with the change, and it seems to be passing for atleast smaller runs (earlier it was failing consistently).

If this result holds up (we will probably see a few categories of bugs still I believe), it suggests that we can likely enable elegant stepdown across our WT testing successfully (once we fix any remaining bugs). 🤞

File: src/conn/conn_layered.c
Comment (luke-pearson) [inline src/conn/conn_layered.c:1751]: I don't think this necessarily needs to be fixed in this PR but I'm not a fan of this function `__disagg_mark_btrees_readonly_then_step_down` in general. The parent function is called `__disagg_step_down`, so why does the step down not take place in the parent? Instead we call into a dual purpose function.

I'm assuming that this semantic is created by the `WT_WITH_HANDLE_LIST_READ_LOCK` where we need to do both of those things inside the same lock and pushed the work down into the function to handle that. The correct choice would be to expand the scope of the lock in this function and perform the step down, the step down is only 2 LOC:
```
    /* Step down to the follower mode. */
    conn->layered_table_manager.leader = false;
    WT_STAT_CONN_SET(session, disagg_role_leader, 0);
```

Then the child function can be `__disagg_mark_btrees_readonly`. Given Dan is owning this I'm not sure if I feel like getting him to fix it, and it expands the scope of the PR. After this PR closes out I'll create a follow on ticket. If anyone has a different opinion here please let me know.

File: src/conn/conn_layered.c [hunk 2]
```diff
@@ -1714,6 +1714,18 @@ __disagg_mark_btrees_readonly_then_step_down(WT_SESSION_IMPL *session)
         /* Mark the disaggregated as readonly. */
         F_SET(btree, WT_BTREE_READONLY);
 
+        /*
+         * Mark the handle outdated so the next leader opens a fresh one rather than reusing this
```
Comment (luke-pearson) [inline src/conn/conn_layered.c]: This comment almost reads backwards to me, we're currently the leader, and marking all of our b-trees as outdated. So why would we have resident follower era-pages? It then goes on to say "Carrying those pages into the leader" but we're already leader so we'd be carrying them into the follower? 

As a side note I think the comment can be improved and shortened on top of my open question.

**Review by luke-pearson — APPROVED** [review]
LGTM.
